plugins/videobasedtracker/HDKLedIdentifier.cpp

```diff
@@ -0,0 +1,135 @@
+        const PatternStringList &PATTERNS) {
+        // Ensure that we have at least one entry in our list and
+        // find the length of the first entry.
+        if (PATTERNS.size() == 0) {
+            d_length = 0;
+            return;
+        }
+        d_length = PATTERNS[0].size();
+
+        // Decode each string into a new list of boolean values, making
+        // sure each have the correct length.
+        for (size_t i = 0; i < PATTERNS.size(); i++) {
+
+            // Make sure the pattern is the correct length.
+            if (PATTERNS[i].size() != d_length) {
+                d_patterns.clear();
+                d_length = 0;
+                return;
+            }
+
+            // Make a new boolean-list encoding from it, replacing every
+            // non-'.' character with true and every '.' with false.
+            std::list<bool> pattern;
+            for (size_t j = 0; j < PATTERNS[i].size(); j++) {
+                if (PATTERNS[i][j] == '.') {
+                    pattern.push_back(false);
+                } else {
+                    pattern.push_back(true);
+                }
+            }
+
+            // Add the pattern to the vector of lists.
+            d_patterns.push_back(pattern);
+        }
+        // std::cout << "XXX d_length = " << d_length << ", num patterns = " <<
+        // d_patterns.size() << std::endl;
+    }
+
+    int OsvrHdkLedIdentifier::getId(std::list<float> brightnesses) const {
+        // If we don't have at least the required number of frames of data, we
+        // don't know anything.
+        if (brightnesses.size() < d_length) {
+            return -1;
+        }
+
+        // We only care about the d_length most-recent levels.
+        truncateBrightnessListTo(brightnesses, d_length);
+
+        // Compute the minimum and maximum brightness values.  If
+        // they are too close to each other, we have a light rather
+        // than an LED.  If not, compute a threshold to separate the
+        // 0's and 1's.
+        auto extrema = findMinMaxBrightness(brightnesses);
+        const auto minVal = extrema.first;
+        const auto maxVal = extrema.second;
+        static const double TODO_MIN_BRIGHTNESS_DIFF = 0.5;
+        if (maxVal - minVal <= TODO_MIN_BRIGHTNESS_DIFF) {
+            return -2;
+        }
+        const auto threshold = (minVal + maxVal) / 2;
+
+        // Get a list of boolean values for 0's and 1's using
+        // the threshold computed above.
+        auto bits = getBitsUsingThreshold(brightnesses, threshold);
+
+        // Search through the available patterns to see if the passed-in
+        // pattern matches any of them.  If so, return that pattern.  We
+        // need to check all potential rotations of the pattern, since we
+        // don't know when the code started.  For the HDK, the codes are
+        // rotationally invariant.
+
+        /// @todo feels like there should be a good algorithm for
```

Comment:
I could certainly believe that there is a more efficient way to do this.  The system was operating at full video rate on my laptop and there are features to add, so I wasn't trying to optimize every step.  Whether the more-efficient way is easy to understand as working properly I don't know, so I recommend holding off until we have regression tests on the subparts.  Does profiling show this to be a large fraction of the overall time?  I would expect the video processing to be taking the lion's share of the time.